src/nc/nc_utils.js

```diff
@@ -24,7 +28,136 @@ function check_root_account_owns_user(root_account, account) {
+        if (use_iam_error) {
+            const { code, http_code, type } = IamError.ValidationError;
+            throw new IamError({ code, message: message_with_details, http_code, type });
+        }
+        throw new Error(message_with_details);
+    }
+}
+
+/**
+ * _length_check_input checks that the input length is between the min and the max value
+ * @param {number} min_length
+ * @param {number} max_length
+ * @param {string} input_value
+ * @param {string} parameter_name
+ * @param {boolean} [use_iam_error]
+ */
+function _length_check_input(min_length, max_length, input_value, parameter_name, use_iam_error = true) {
+    _length_min_check_input(min_length, input_value, parameter_name, use_iam_error);
+    _length_max_check_input(max_length, input_value, parameter_name, use_iam_error);
+}
+
+/**
+ * _length_min_check_input checks if the input is lower than the min length
+ * @param {number} min_length
+ * @param {any} input_value
+ * @param {string} parameter_name
+ * @param {boolean} [use_iam_error]
+ */
+function _length_min_check_input(min_length, input_value, parameter_name, use_iam_error = true) {
+    const input_length = input_value.length;
+    if (input_length < min_length) {
+        const message_with_details = `Invalid length for parameter ${parameter_name}, ` +
+            `value: ${input_length}, valid min length: ${min_length}`;
+        if (use_iam_error) {
+            const { code, http_code, type } = IamError.ValidationError;
+            throw new IamError({ code, message: message_with_details, http_code, type });
+        }
+        throw new Error(message_with_details);
+    }
+}
+
+/**
+ * _length_max_check_input checks if the input is higher than the max length
+ * @param {number} max_length
+ * @param {any} input_value
+ * @param {string} parameter_name
+ * @param {boolean} [use_iam_error]
+ */
+function _length_max_check_input(max_length, input_value, parameter_name, use_iam_error = true) {
+    const input_length = input_value.length;
+    if (input_length > max_length) {
+        const message_with_details = `1 validation error detected: Value ${input_value} at ` +
+            `'${parameter_name}' failed to satisfy constraint:` +
+            `Member must have length less than or equal to ${max_length}`;
+        if (use_iam_error) {
+            const { code, http_code, type } = IamError.ValidationError;
+            throw new IamError({ code, message: message_with_details, http_code, type });
+        }
+        throw new Error(message_with_details);
+    }
+}
+
+/**
+ * validate_username will validate:
+ * 1. type
+ * 2. length
+ * 3. regex (from AWS docs)
+ * 4. additional internal restrictions
+ * @param {string} input_username
+ * @param {string} parameter_name
+ * @param {boolean} [use_iam_error]
+ */
+function validate_username(input_username, parameter_name = iam_constants.IAM_PARAMETER_NAME.USERNAME, use_iam_error = true) {
```

Comment:
if you remember our discussion about Internal vs external errors, use_iam_error is the probably the exact reason why we used RPC internal errors and then converted them to external errors. why not consider using RPC errors?

Reply:
I don't know if it's small or not, I think using `use_iam_error = true` is not designed well